File: bindings/java/src/main/java/systems/zlink/contracts/sockets/RoutedSocketContracts/RouterSocket.java
/* SPDX-License-Identifier: MPL-2.0 */

package systems.zlink.contracts.sockets;

import java.util.List;
import systems.zlink.contracts.core.RoutingId;
import systems.zlink.contracts.messaging.Received;
import systems.zlink.contracts.messaging.ReplyOperation;
import systems.zlink.contracts.messaging.ReplyToken;
import systems.zlink.contracts.messaging.RequestOperation;
import systems.zlink.contracts.messaging.SendOperation;

/** Routes messages to peers addressed by routing id; the request/reply server side. */
public interface RouterSocket extends Socket {
    void bind(String endpoint);
    void connect(String endpoint);
    void unbind(String endpoint);
    void disconnect(String endpoint);
    void disconnectRid(RoutingId routingId);
    void setRoutingId(RoutingId rid);
    RoutingId getRoutingId();

    /**
     * Returns every selected route atomically, one row per routing id. A
     * successful call clears {@code POLLROUTE} readiness unless a later change
     * raced with it. Keep one route observer per socket.
     */
    List<RouterRoute> routesSnapshot();
    SendOperation send(RoutingId rid);
    boolean recv(Received result, RecvFlags flags);
    RequestOperation request(RoutingId rid);
    ReplyOperation reply(RoutingId rid, ReplyToken token);
    @Override RouterSocketOptions options();
}

File: bindings/node/src/zlink/contracts/sockets/router_socket.ts
// SPDX-License-Identifier: MPL-2.0

import type { RoutingId } from '../core';
import type { Received, ReplyToken } from '../messaging';
import type {
  ReplyOperation,
  RequestOperation,
  SendOperation,
} from '../messaging';
import type { RecvFlags } from './socket_constants';
import type { RouterSocketOptions } from './socket_options';
import type { ConnectableSocket } from './socket';

/**
 * ROUTER socket: routes messages to peers addressed by routing id, the server
 * side of asynchronous request/reply.
 */
/**
 * One row of the ROUTER selected-route snapshot. Core selects one application
 * route per routing id; `routeGeneration` is a nonzero opaque token that
 * changes whenever that selection changes and is compared only for equality.
 */
export interface RouterRoute {
  readonly routingId: RoutingId;
  readonly routeGeneration: bigint;
}

export interface RouterSocket extends ConnectableSocket {
  /** The ROUTER-specific typed options facade. */
  readonly options: RouterSocketOptions;
  /** Begin a send addressed to `routingId`; parts are consumed on a successful submit. */
  send(routingId: RoutingId): SendOperation;
  /**
   * Receive a routed message into `result`; false when `RecvFlags.DontWait` is
   * set and no message is available.
   */
  recv(result: Received, flags?: RecvFlags): boolean;
  /**
   * Set the routing id that identifies this ROUTER to its peers. Apply before
   * connecting so peers observe it from the first message.
   */
  setRoutingId(routingId: RoutingId): void;
  /** Return the routing id that identifies this ROUTER to its peers. */
  getRoutingId(): RoutingId;
  /** Begin a request to peer `peerRid`; parts are consumed on submit and a reply is awaited. */
  request(peerRid: RoutingId): RequestOperation;
  /** Begin a reply using the opaque token returned by this ROUTER's request receive. */
  reply(peerRid: RoutingId, token: ReplyToken): ReplyOperation;
  /**
   * Return the selected route of every routing id atomically. A routing id
   * without a row has no selected route. A successful snapshot clears
   * `PollEventFlag.PollRoute` readiness; one observer per socket calls it.
   */
  routesSnapshot(): RouterRoute[];
}
